controller/appcontroller.go

```diff
@@ -1535,6 +1604,53 @@ func currentSourceEqualsSyncedSource(app *appv1.Application) bool {
 	return app.Spec.Source.Equals(&app.Status.Sync.ComparedTo.Source)
 }
 
+// shouldRefreshForDependency returns whether an app should be refreshed for a
+// change in the state of one of its dependencies.
+func (ctrl *ApplicationController) shouldRefreshForDependency(app *appv1.Application) bool {
+	logCtx := log.WithField("application", app.QualifiedName())
+
+	// No need to refresh when we're not waiting for dependencies to sync
+	if app.Spec.DependsOn == nil || !app.IsWaiting() {
+		logCtx.Debugf("not waiting for dependencies, skipping refresh")
+		return false
+	}
+
+	// If we're waiting for any dependency to be created, we need refresh to
+	// see whether they're created by now.
+	if app.Status.OperationState.BlockedOnEmpty {
+		return true
+	}
```

Comment:
If there are no ApplicationSelectors specified, it means the Application does not have a dependency on anything. In that case, we should check if the Application has any ApplicationSelectors specified along with `BlockedOnEmpty`

Reply:
The check for that is just above this condition you are highlighting here